File: Java/src/Act1/Biblioteca.java
package Act1;

import java.util.ArrayList;
import java.util.List;

public class Biblioteca {
    private List<Libro> libros;
    private List<Usuario> usuarios;

    public Biblioteca() {
        libros = new ArrayList<>();
        usuarios = new ArrayList<>();
    }

    public void agregarLibro(Libro libro) {
        libros.add(libro);
        System.out.println("Libro agregado: " + libro.getTitulo());
    }

    public void registrarUsuario(Usuario usuario) {
        usuarios.add(usuario);
        System.out.println("Usuario registrado: " + usuario.getNombre());
    }

    public void prestarLibro(String titulo, int usuarioId) {
        Libro libro = buscarLibro(titulo);
        Usuario usuario = buscarUsuario(usuarioId);

        if (libro == null) {
            System.out.println("El libro " + titulo + " no existe en la biblioteca.");
            return;
        }

        if (usuario == null) {
            System.out.println("El usuario con ID " + usuarioId + " no está registrado.");
            return;
        }

        if (libro.isPrestado()) {
            System.out.println("El libro " + titulo + " ya está prestado.");
        } else {
            libro.prestar();
            System.out.println("El libro " + titulo + " ha sido prestado a " + usuario.getNombre());
        }
    }

    public void devolverLibro(String titulo) {
        Libro libro = buscarLibro(titulo);

        if (libro == null) {
            System.out.println("El libro " + titulo + " no existe en la biblioteca.");
            return;
        }

        if (!libro.isPrestado()) {
            System.out.println("El libro " + titulo + " no está prestado.");
        } else {
            libro.devolver();
            System.out.println("El libro " + titulo + " ha sido devuelto.");
        }
    }

    private Libro buscarLibro(String titulo) {
        for (Libro libro : libros) {
            if (libro.getTitulo().equalsIgnoreCase(titulo)) {
                return libro;
            }
        }
        return null;
    }

    private Usuario buscarUsuario(int id) {
        for (Usuario usuario : usuarios) {
            if (usuario.getId() == id) {
                return usuario;
            }
        }
        return null;
    }
}

File: KotlinCode/src/Act1/Biblioteca.kt
package Act1

import java.util.ArrayList

class Biblioteca {
    private val libros: MutableList<Libro>
    private val usuarios: MutableList<Usuario>

    init {
        libros = ArrayList<Libro>()
        usuarios = ArrayList<Usuario>()
    }

    fun agregarLibro(libro: Libro) {
        libros.add(libro)
        println("Libro agregado: " + libro.titulo)
    }

    fun registrarUsuario(usuario: Usuario) {
        usuarios.add(usuario)
        println("Usuario registrado: " + usuario.nombre)
    }

    fun prestarLibro(titulo: String?, usuarioId: Int) {
        val libro = buscarLibro(titulo)
        val usuario = buscarUsuario(usuarioId)

        if (libro == null) {
            println("El libro " + titulo + " no existe en la biblioteca.")
            return
        }

        if (usuario == null) {
            println("El usuario con ID " + usuarioId + " no está registrado.")
            return
        }

        if (libro.isPrestado()) {
            println("El libro " + titulo + " ya está prestado.")
        } else {
            libro.prestar()
            println("El libro " + titulo + " ha sido prestado a " + usuario.nombre)
        }
    }

    fun devolverLibro(titulo: String?) {
        val libro = buscarLibro(titulo)

        if (libro == null) {
            println("El libro " + titulo + " no existe en la biblioteca.")
            return
        }

        if (!libro.isPrestado()) {
            println("El libro " + titulo + " no está prestado.")
        } else {
            libro.devolver()
            println("El libro " + titulo + " ha sido devuelto.")
        }
    }

    private fun buscarLibro(titulo: String?): Libro? {
        for (libro in libros) {
            if (libro.titulo.equals(titulo, ignoreCase = true)) {
                return libro
            }
        }
        return null
    }

    private fun buscarUsuario(id: Int): Usuario? {
        for (usuario in usuarios) {
            if (usuario.id == id) {
                return usuario
            }
        }
        return null
    }
}

